azurelinuxagent/ga/agent_update_handler.py

```diff
@@ -145,30 +143,34 @@ def run(self, goal_state, ext_gs_updated):
             if not conf.get_autoupdate_enabled() or not conf.get_download_new_agents():
                 return
 
-            # verify if agent update is allowed this time (RSM checks new goal state; self-update checks manifest download interval)
-            if not self._updater.is_update_allowed_this_time(ext_gs_updated):
-                return
-
             self._gs_id = goal_state.extensions_goal_state.id
-            agent_family = self._get_agent_family_manifest(goal_state)
+            agent_family = self._get_agent_family_manifest(goal_state, ext_gs_updated)
 
-            # updater will return RSM enabled or disabled if we need to switch to self-update or rsm update
-            updater_mode = self._updater.check_and_switch_updater_if_changed(agent_family, self._gs_id, ext_gs_updated)
+            # Updater will return True or False if we need to switch the updater
+            # If self-update receives RSM update request, it will switch to RSM updater
+            # If RSM updater stopped receiving RSM request, it will switch to self-update
+            # No change in updater if GS not updated
+            switch_to_rsm_updater = self._updater.is_gs_requested_rsm_update(agent_family, self._gs_id, ext_gs_updated)
 
-            if updater_mode == RSMUpdates.Disabled:
+            if not switch_to_rsm_updater and isinstance(self._updater, RSMVersionUpdater):
```

Comment:
This check confused me at first because the variable name says 'switch'. The value could be true, but we're not doing a switch because the updater already is RSMUpdater. I think 'is_RSM_update' is clearer/more readable

Reply:
I also find this code confusing, but the other way around: if one looks at the documentation for this method, it begins with "Checks if there is a new goal state and decide if we need to continue with rsm update or switch to self-update.", which makes me think both the method and the variable should be renamed to something like "should_switch_to_self_update", but then one continues reading the documentation and the implementation of the method and it looks like it is whether the machine is enabled for RSM updates, which seems to be confirmed by the message we log just in one line below. In that case, probably the method and variable should be renamed to "is_rsm_update_enabled".

Another issue with that method is that it looks like a query, just a check on state, but it actually updates the state (self._gs_id = gs_id) as a side effect.